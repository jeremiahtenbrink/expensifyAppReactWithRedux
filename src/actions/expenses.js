import uuid from "uuid";

//ADD EXPENSE ACTION
export const addExpense = (
	{
		description = "",
		note = "",
		amount = 0,
		createdAt = 0,
	} = {} ) => ( {
	type   : "ADD EXPENSE",
	expense: {
		id: uuid(),
		description,
		note,
		amount,
		createdAt,
	},
	
} );

//EDIT EXPENSE ACTION
export const editExpense = (
	{
		id,
		updates,
	} = {} ) => ( {
	type: "EDIT EXPENSE",
	id,
	updates,
} );

// REMOVE EXPENSE BY ID
export const removeExpense = ( { id = undefined } = {} ) => ( {
	type: "REMOVE EXPENSE",
	id,
} );
