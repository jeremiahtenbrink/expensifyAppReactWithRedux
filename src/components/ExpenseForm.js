import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import { connect } from "react-redux";

export class ExpenseForm extends React.Component {
	constructor ( props )
	{
		super( props );
		
		this.state = {
			description    : props.expense ? props.expense.description : "",
			note           : props.expense ? props.expense.note : "",
			amount         : props.expense ? ( props.expense.amount / 100 ).toString() : "",
			createdAt      : props.expense ? moment( props.expense.createdAt ) : moment(),
			calendarFocused: false,
			errorMessage   : "",
		};
	}
	
	onDescriptionChange = ( e ) => {
		const description = e.target.value;
		this.setState( () => ( { description } ) );
	};
	
	onTextAreaChange = ( e ) => {
		const note = e.target.value;
		this.setState( () => ( { note } ) );
	};
	
	onAmountChange = ( e ) => {
		const amount = e.target.value;
		if ( !amount || amount.match( /^\d{1,}(\.\d{0,2})?$/ ) ){
			this.setState( () => ( { amount } ) );
		}
	};
	
	onDateChange = ( createdAt ) => {
		if ( createdAt ){
			this.setState( () => ( { createdAt } ) );
		}
	};
	
	onFocuseChange = ( { focused } ) => {
		this.setState( () => ( { calendarFocused: focused } ) );
	};
	
	onSubmit = ( e ) => {
		e.preventDefault();
		
		if ( !this.state.description || !this.state.amount ){
			// set error state
			this.setState( () => ( { errorMessage: "Please provide description and amount." } ) );
		} else{
			this.setState( () => ( { errorMessage: "" } ) );
			// send to edit or add expense page
			this.props.onSubmit( {
				description: this.state.description,
				amount     : parseFloat( this.state.amount ) * 100,
				createdAt  : this.state.createdAt.valueOf(),
				note       : this.state.note,
			} );
		}
	};
	
	render ()
	{
		return (
			<div>
				{ this.state.errorMessage && <p>{ this.state.errorMessage }</p> }
				<form onSubmit={ this.onSubmit }>
					<input type="text"
					       placeholder="Description"
					       autoFocus
					       value={ this.state.description }
					       onChange={ this.onDescriptionChange }
					/>
					<input type="text"
					       placeholder="Amount"
					       value={ this.state.amount }
					       onChange={ this.onAmountChange }
					/>
					<SingleDatePicker
						date={ this.state.createdAt }
						onDateChange={ this.onDateChange }
						focused={ this.state.calendarFocused }
						onFocusChange={ this.onFocuseChange }
						numberOfMonths={ 1 }
						isOutsideRange={ () => false }
					/>
					<textarea name="note" id="" cols="20" rows="2"
					          value={ this.state.note }
					          placeholder="Add a note for your expense"
					          onChange={ this.onTextAreaChange }/>
					<button>Add Expense</button>
				
				</form>
			</div>
		);
	}
}

export default connect()( ExpenseForm );
