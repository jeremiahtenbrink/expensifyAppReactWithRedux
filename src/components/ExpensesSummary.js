import React from "react";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import { getExpensesTotal } from "../selectors/expenses-total";
import { connect} from "react-redux";

export const ExpensesSummary = ( { expenseCount, expensesTotal, totalNumberOfExpenses } ) => {
	
	const expenseWord = expenseCount === 1 ? "expense" : "expenses";
	const formattedExpensesTotal = numeral( expensesTotal / 100 ).format( "$0,0.00" );
	
	return (
		<div>
			<h3>
				Showing { expenseCount } of { totalNumberOfExpenses } { expenseWord }:
				
			</h3>
			<h3>
				Totalling: { formattedExpensesTotal }
			</h3>
		</div>
	);
};

const mapStateToProps = ( state ) => {
	const visibleExpenses = selectExpenses( state.expenses, state.filter );
	
	return {
		expenseCount        : visibleExpenses.length === 0 ? 0 : visibleExpenses.length,
		expensesTotal        : getExpensesTotal( visibleExpenses ),
		totalNumberOfExpenses: state.expenses.length,
	};
};

export default connect( mapStateToProps )( ExpensesSummary );


