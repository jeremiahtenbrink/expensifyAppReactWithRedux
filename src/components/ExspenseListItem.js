import React from "react";
import { Link } from "react-router-dom";

const ExspenseListItem = ( { description, amount, createdAt, id } ) => {
	return (
		<div>
			<h3><Link to={ "/edit/" + id }>{ description }</Link></h3>
			<p>{ amount } -- { createdAt }</p>
		
		</div>
	);
};

export default ExspenseListItem;
ExspenseListItem.propTypes = {};
ExspenseListItem.defaultProps = {};

