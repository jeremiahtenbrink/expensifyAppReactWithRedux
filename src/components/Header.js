import { NavLink } from "react-router-dom";
import React from "react";

const Header = () => (
	<header>
		<h1>Expensify</h1>
		<nav>
			<ul>
				<li><NavLink to="/" activeClassName="is-active" exact={ true }>Dashboard</NavLink></li>
				<li><NavLink to="/create" activeClassName="is-active">Create Expense</NavLink></li>u
			</ul>
		</nav>
	</header>
);

export default Header;