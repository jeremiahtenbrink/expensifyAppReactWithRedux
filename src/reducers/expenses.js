const expensesReducerDefaultState = [];

export default ( state = expensesReducerDefaultState, action ) => {
	switch ( action.type ){
		case "ADD EXPENSE":
			return [
				...state,
				action.expense,
			];
		case "REMOVE EXPENSE":
			return state.filter( ( expense ) => expense.id !== action.id );
		case "EDIT EXPENSE":
			return state.map( ( expense ) => {
				if ( expense.id === action.id ){
					const newExpense = { ...expense, ...action.updates };
					
					return newExpense;
					
				} else{
					return expense;
				}
			} );
		default:
			return state;
	}
};

