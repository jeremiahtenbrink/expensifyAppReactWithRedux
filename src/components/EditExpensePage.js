import React, { Component } from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends Component {
	
	onEditExpense = (expense) => {
		this.props.editExpense( {
			id: this.props.expense.id,
			updates: expense} );
		this.props.history.push( "/" );
	};
	
	onRemoveExpense = () => {
		
		this.props.removeExpense({ id: this.props.expense.id });
		this.props.history.push( "/" );
	};
	
	render ()
	{
		
		return (
			<div>
				<ExpenseForm id="expense-form"
					expense={ this.props.expense }
					onSubmit={this.onEditExpense}
				/>
				
				<button onClick={this.onRemoveExpense}>Remove
				</button>
			</div>
		
		);
	}
}

const mapStateToProps = ( state, props ) => {
	return {
		expense: state.expenses.find( ( expense ) => expense.id === props.match.params.id ),
	};
};

const mapDispatchToProps = (dispatch) => ({
	editExpense: (expense) => dispatch(editExpense(expense)),
	removeExpense: (idObject) => dispatch(removeExpense( idObject ))
});

export default connect( mapStateToProps, mapDispatchToProps )( EditExpensePage );