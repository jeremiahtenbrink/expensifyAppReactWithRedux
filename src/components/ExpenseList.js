import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import ExspenseListItem from "./ExspenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = ( props ) => (
	<div>
		{
			props.expenses.length === 0 ? (
				<p>No expenses</p>
			): (
				props.expenses.map( ( expense ) => {
					return <ExspenseListItem key={ expense.id } { ...expense } />;
				} )
			)
		}
		
	</div>
);

const mapStateToProps = state => ( {
	expenses: selectExpenses( state.expenses, state.filter ),
} );

export default connect( mapStateToProps )( ExpenseList );

ExpenseList.propTypes = {};
ExpenseList.defaultProps = {};


