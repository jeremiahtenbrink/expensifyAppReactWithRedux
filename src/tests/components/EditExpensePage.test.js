import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let editExpense, removeExpense, history, wrapper;

beforeEach( () => {
	editExpense = jest.fn();
	removeExpense = jest.fn();
	history = { push: jest.fn() };
	wrapper =
		shallow( <EditExpensePage expense={ expenses[ 1 ] } editExpense={ editExpense } removeExpense={ removeExpense }
		                          history={ history }/> );
} );

test( "should render edit expense page correctly", () => {
	expect( wrapper ).toMatchSnapshot();
} );


test( "should handle onEditExpense", () => {
	wrapper.find( '#expense-form' ).prop( "onSubmit" )( expenses[ 1 ] );
	expect( history.push ).toHaveBeenLastCalledWith( "/" );
	expect( editExpense ).toHaveBeenLastCalledWith( {
		id: "2",
		updates: expenses[ 1 ]
	});
} );

test( "should handle onRemoveExpense", () => {
	wrapper.find( 'button' ).prop( "onClick" )();
	expect( history.push ).toHaveBeenLastCalledWith( "/" );
	expect( removeExpense ).toHaveBeenLastCalledWith( {
		id: "2",
	});
} );

