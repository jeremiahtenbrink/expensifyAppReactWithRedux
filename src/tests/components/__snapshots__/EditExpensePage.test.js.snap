// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page correctly 1`] = `
<div>
  <Connect(ExpenseForm)
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    id="expense-form"
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
